src/omerocrate/uploader.py

```diff
@@ -21,6 +23,95 @@
 Namespaces = dict[str, URIRef]
 Variables = dict[str, Identifier]
 
+
+class SegmentationUploader(BaseModel, arbitrary_types_allowed=True):
+    """
+    Class that handles uploading of segmentation masks to OMERO.
+    Users can subclass this to customise segmentation upload behavior.
+    """
+    conn: gateway.BlitzGateway
+    "OMERO connection object, typically obtained using [`from_env`][omerocrate.gateway.from_env]"
+
+    def process_segmentation(self, segmentation_path: Path, image: gateway.ImageWrapper) -> None:
+        """
+        Load segmentation mask and upload to OMERO for the given image URI.
+        Users should override this method to implement segmentation upload logic.
+        """
+        raise NotImplementedError("process_segmentation() must be implemented in a subclass")
+
+
+class OmeNgffUploader(SegmentationUploader):
+    """
+    Subclass of SegmentationUploader that uses Glencoe's ROI_Converter_NGFF to upload segmentations.
+    Assumes that the segmentation file is a CSV with a 'polygon' or 'geometry' column containing WKT
+    polygons, and an 'object' or 'id' column for object identifiers.
+    Requires ROI_Converter_NGFF to be installed and accessible in the Python environment.
+    """
+
+    def __init__(self, conn: gateway.BlitzGateway):
+        if importlib.util.find_spec("ROI_Converter_NGFF") is None:
+            raise ImportError("ROI_Converter_NGFF is required for OmeNgffUploader.")
+        super().__init__(conn=conn)
+
+    def process_segmentation(self, segmentation_path: Path, image: gateway.ImageWrapper) -> None:
+        """
+        Load segmentation mask and upload to OMERO for the given image URI.
+        By default, uses Glencoe's ROI_Converter_NGFF to parse file, rasterise shapes
+        into a zarr file, and register the mask with OMERO.
+        """
+        header: list[str] = []
+        with open(segmentation_path, 'r') as f:
+            header = f.readline().strip().split(',')
+
+        # Check geometry column name
+        geometry_column = None
+        if "polygon" in header:
+            geometry_column = "polygon"
+        elif "geometry" in header:
+            geometry_column = "geometry"
+        if not geometry_column:
+            raise ValueError(
+                f"Segmentation file {segmentation_path} does not contain a "
+                f"'{geometry_column}' or 'polygon' column"
+            )
+
+        from ROI_Converter_NGFF import raster
+        args = {
+            "input_file": str(segmentation_path),
+            "register_to": image.getId(),
+            "name": segmentation_path.stem,
+            "directory": os.getenv("UPLOAD_DIRECTORY", None),
```

Comment:
Can you make the directory into a model field? Or can you explain why this is better as an environment variable? Is this the location on the OMERO server where the segmentation will live?

Reply:
Good catch, the env was just meant to be a placeholder. I've made it into a model field and settable via the CLI.